examples/docs_projects/project_components_pdf_extraction/pdf_extraction/lib/pdf_extraction_resource.py

```diff
@@ -0,0 +1,467 @@
+    """Initialize the PDF processor resource.
+
+    Args:
+        language: OCR language (default: 'eng')
+        dpi: DPI for PDF conversion (default: 300)
+        openai_api_key: OpenAI API key for validation (optional)
+        preprocess: Whether to preprocess images (default: True)
+    """
+
+    language: str
+    dpi: int
+    openai_api_key: str
+    preprocess: bool = True
+    log: ClassVar = dg.get_dagster_logger()  # Add type annotation with ClassVar
+    output_dir: str
+
+    @property
+    def openai_headers(self):
+        return {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {self.openai_api_key}",
+        }
+
+    def preprocess_image(self, image_path: str, save_preprocessed: bool = True) -> tuple:
+        """Preprocess image to improve OCR quality.
+
+        Args:
+            image_path: Path to the image file
+            save_preprocessed: Whether to save the preprocessed image
+
+        Returns:
+            Tuple of (preprocessed image, path to saved preprocessed image)
+        """
+        self.log.info(f"Preprocessing image: {image_path}")
+
+        # Load the image
+        image = Image.open(image_path)
+
+        # Convert to grayscale
+        image = image.convert("L")
+
+        # 1. Increase contrast
+        enhancer = ImageEnhance.Contrast(image)
+        image = enhancer.enhance(2.0)  # Increase contrast by factor of 2
+
+        # 2. Apply threshold
+        threshold = 200  # Threshold value (0-255)
+        image = image.point(lambda p: 255 if p > threshold else 0)
+
+        # 3. Remove small noise
+        image = image.filter(ImageFilter.MedianFilter(size=3))
+
+        # 4. Sharpen the image
+        image = image.filter(ImageFilter.SHARPEN)
+        image = image.filter(ImageFilter.SHARPEN)  # Apply twice for stronger effect
+
+        # 5. Enhance edges
+        image = image.filter(ImageFilter.EDGE_ENHANCE)
+
+        # Save preprocessed image if requested
+        if save_preprocessed:
+            # Generate preprocessed image path
+            image_dir = os.path.dirname(image_path)
+            image_filename = os.path.basename(image_path)
+            preprocessed_path = os.path.join(image_dir, f"preprocessed_{image_filename}")
+
+            # Save the preprocessed image
+            image.save(preprocessed_path)
+            self.log.info(f"Saved preprocessed image to {preprocessed_path}")
+
+            return image, preprocessed_path
+
+        return image, None
+
+    def convert_pdf_to_images(
+        self, pdf_path: str, output_folder: Optional[str] = None, pages: Optional[list[int]] = None
+    ) -> dict[str, Any]:
+        """Convert PDF to images and save them to a folder.
+
+        Args:
+            pdf_path: Path to the PDF file
+            output_folder: Custom output folder path. If None, creates folder with PDF name
+            pages: List of page numbers to extract (0-based). If None, extract all pages
```

Comment:
Found this `pages` parameter a little confusing. At first I read it as being an arbitrary set of numbers [2,4,5,7,9] but it seems like it is more of a range 3-7. I think the code could be a little more clear and cleaned up.

If it arbitrary you could just iterate through the list and `continue` for any values not in the `pages` set rather than using the `page_num` and `page_number` parsing.

If it is start and end value, I would include those as separate input parameters and then just slice the images list like `images[start_page_num:end_page_num]`